### PR TITLE
Feature/search 703

### PR DESCRIPTION
**Pull request by joel-bernstein**

### CODE REVIEW

**Comment by joel-bernstein** [comment]
I believe this is ready for review. There is still some System.out debugging in the code which I will remove shortly.

File: alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java [hunk 1]
```diff
@@ -838,6 +843,153 @@ private void actionREPORT(SolrQueryResponse rsp, SolrParams params, String cname
...
+            long endRange = dbidRangeRouter.getEndRange();
+
+            long maxNodeId = informationServer.maxNodeId();
+            long nodeCount = informationServer.nodeCount();
+
+            long bestGuess = -1;  // -1 means expansion cannot be done. Either because expansion
+                                 // has already happened or we're above safe range
+
+            long range = endRange - startRange; // We want this many nodes on the server
+            //System.out.println("##### Range:" + range);
```
Comment (andyhind) [inline alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java]: Remove System.out or add as debug loggin

File: alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java [hunk 2]
```diff
@@ -838,6 +843,153 @@ private void actionREPORT(SolrQueryResponse rsp, SolrParams params, String cname
...
+                    {
+                        if(density >= 1)
+                        {
+                            //This is fully dense shard. I'm not sure if it's possible to have more nodes on the shards
+                            //then the offset, but if it does happen don't expand.
+                            bestGuess=0;
+                        }
+                        else
+                        {
+                            double multiplier = range / (density*range); // Extrapolate across the entire range to find the multiplier
```
Comment (andyhind) [inline alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java]: multiplier = 1.0 / density;

File: alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java [hunk 3]
```diff
@@ -838,6 +843,153 @@ private void actionREPORT(SolrQueryResponse rsp, SolrParams params, String cname
...
+                        bestGuess = 0; // We're below the midpoint so it's to early to make a guess.
+                    }
+                }
+            }
+
+            rsp.add("start", startRange);
+            rsp.add("end", endRange);
+            rsp.add("nodeCount", nodeCount);
+            rsp.add("maxDbid", maxNodeId);
+            rsp.add("density", Math.abs(density));
```
Comment (andyhind) [inline alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java]: Density must be +ve unless I have missed something.

The calculation is correct for the value to add to the current max value to get the desired number of nodes in the shard.

File: alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java [hunk 4]
```diff
@@ -838,6 +843,153 @@ private void actionREPORT(SolrQueryResponse rsp, SolrParams params, String cname
...
+        MetadataTracker metadataTracker = null;
+        for(Tracker tracker : trackers)
+        {
+            if(tracker instanceof MetadataTracker)
+            {
+                metadataTracker = (MetadataTracker)tracker;
+            }
+        }
+
+        DocRouter docRouter = metadataTracker.getDocRouter();
```
Comment (andyhind) [inline alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java]: Possible but unlikely null metadataTracker

File: alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java [hunk 5]
```diff
@@ -838,6 +843,153 @@ private void actionREPORT(SolrQueryResponse rsp, SolrParams params, String cname
...
+            rsp.add("exception", "ERROR: Wrong document router type:"+docRouter.getClass().getSimpleName());
+            return;
+        }
+    }
+
+    private synchronized void expand(SolrQueryResponse rsp, SolrParams params, String cname)
+        throws IOException
+    {
+        InformationServer informationServer = informationServers.get(cname);
+        //Get the in-memory information from MetaDataTracker
```
Comment (andyhind) [inline alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java]: 940-952 is common

File: alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java [hunk 6]
```diff
@@ -1674,6 +1657,88 @@ public void putAclTransactionState(UpdateRequestProcessor processor, SolrQueryRe
         }
     }
 
+    public void capIndex(long dbid) throws IOException
```
Comment (andyhind) [inline alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java]: Missing try finally for request and processor tidy up

File: alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java [hunk 7]
```diff
@@ -1674,6 +1657,88 @@ public void putAclTransactionState(UpdateRequestProcessor processor, SolrQueryRe
...
+    {
+        SolrQueryRequest request = getLocalSolrQueryRequest();
+        UpdateRequestProcessor processor = this.core.getUpdateProcessingChain(null).createProcessor(request, new SolrQueryResponse());
+        AddUpdateCommand cmd = new AddUpdateCommand(request);
+        cmd.overwrite = true;
+        SolrInputDocument input = new SolrInputDocument();
+        input.addField(FIELD_SOLR4_ID, "TRACKER!STATE!CAP");
+        input.addField(FIELD_VERSION, 0);
+        input.addField(FIELD_DBID, -dbid); //Making this negative to ensure it is never confused with node DBID
```
Comment (andyhind) [inline alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java]: Could confuse unbounded lower range if we add range support for DBID ...

File: alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java [hunk 8]
```diff
@@ -1674,6 +1657,88 @@ public void putAclTransactionState(UpdateRequestProcessor processor, SolrQueryRe
...
+    }
+
+    public long getIndexCap()
+    {
+        SolrQueryRequest request = null;
+        try
+        {
+            request = this.getLocalSolrQueryRequest();
+            ModifiableSolrParams params = new ModifiableSolrParams(request.getParams());
+            params.set("q", FIELD_SOLR4_ID+":TRACKER!STATE!CAP");
```
Comment (andyhind) [inline alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java]: Static for .... TRACKER!STATE!CAP

File: alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java [hunk 9]
```diff
@@ -1674,6 +1657,88 @@ public void putAclTransactionState(UpdateRequestProcessor processor, SolrQueryRe
...
+            // Sets the rows to zero, because we actually just want the count
```
Comment (andyhind) [inline alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java]: Misleading comment!

File: alfresco-search/src/main/java/org/alfresco/solr/tracker/MetadataTracker.java [hunk 10]
```diff
@@ -152,6 +157,19 @@ private void trackRepository() throws IOException, AuthenticationException, JSON
             checkRepoAndIndexConsistency(state);
         }
 
+        if(docRouter instanceof DBIDRangeRouter)
```
Comment (andyhind) [inline alfresco-search/src/main/java/org/alfresco/solr/tracker/MetadataTracker.java]: There is an unlikely timing issue here.
Expansion is registered  as part of tracking. At start up you could expand before this is called.
I am not sure where this should live but it should get done once

File: alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java (hunk 9 above)
Comment (joel-bernstein) [reply alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java]: Yes, this is misleading.

File: alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java (hunk 8 above)
Comment (joel-bernstein) [reply alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java]: Will do.

File: alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java (hunk 7 above)
Comment (joel-bernstein) [reply alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java]: Can you elaborate on this?

File: alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java (hunk 6 above)
Comment (joel-bernstein) [reply alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java]: hmm, that's quite bad. I'll fix.

File: alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java (hunk 5 above)
Comment (joel-bernstein) [reply alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java]: Ok, I'll add a method for this.

File: alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java (hunk 2 above)
Comment (joel-bernstein) [reply alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java]: Yes, this is simpler. Is what I have equivalent? I'll change and see what happens with the tests. If the test break then they are not equivalent.

File: alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java (hunk 3 above)
Comment (joel-bernstein) [reply alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java]: Not sure I understand this comment. I'll review the code again.

File: alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java (hunk 1 above)
Comment (joel-bernstein) [reply alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java]: I'll remove all the commented System.out's in the final commit.

File: alfresco-search/src/main/java/org/alfresco/solr/tracker/MetadataTracker.java (hunk 10 above)
Comment (joel-bernstein) [reply alfresco-search/src/main/java/org/alfresco/solr/tracker/MetadataTracker.java]: I ran into this during manual testing. I'll think about an approach for dealing with this in the short term.

File: alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java (hunk 2 above)
Comment (joel-bernstein) [reply alfresco-search/src/main/java/org/alfresco/solr/AlfrescoCoreAdminHandler.java]: Tests still pass after making this change so they were equivalent. I'll go with the suggested formula though as its simpler.

**Pull request merged**